### PR TITLE
Feature/needle in haystack task 2

### PR DESCRIPTION
**Pull request by jixy2012**
# Create Needle in Haystack 
- you can specify a needle in haystack (gittables) dataset along with other eval datasets.
- raise error if only needle in haystack is included and no other datasets. 
- input examples:
    - you can do `TARGET(("Table Retrieval Task", ["fetaqa", "gittables"]))`
    - but CANNOT do `TARGET(("Table Retrieval Task", ["gittables"]))`
- you can also specify how many tables you need from the nih datasets. you can use the `num_tables` field. for example:
```python
from target_benchmark.dataset_loaders.TargetDatasetConfig import (
    DEFAULT_FETAQA_DATASET_CONFIG,
    DEFAULT_GITTABLES_DATASET_CONFIG,
)
from target_benchmark.evaluators import TARGET
from target_benchmark.tasks import TableRetrievalTask

gittables_config = DEFAULT_GITTABLES_DATASET_CONFIG.model_copy()
        gittables_config.num_tables = num_tables
TARGET(
            TableRetrievalTask(
                {
                    "fetaqa": DEFAULT_FETAQA_DATASET_CONFIG,
                    "gittables": gittables_config,
                }
            )
        )
```
- I've added a new issue [here](https://github.com/target-benchmark/target/issues/33#issue-2639650516) to address some issues i have with NIH right now, but want to get this pr done first and improve in the future.

### CODE REVIEW

File: target_benchmark/dataset_loaders/utils.py [hunk 1]
```diff
@@ -157,11 +171,7 @@ def convert_nested_list_to(
         return array_of_arrays_to_df(nested_list)
 
 
-def get_dummy_table_of_format(
-    expected_format: Literal[
-        "array", "nested array", "pandas", "dataframe"
-    ] = "nested array"
-):
+def get_dummy_table_of_format(expected_format: Literal["array", "nested array", "pandas", "dataframe"] = "nested array"):
```
Comment (madelonhulsebos) [inline target_benchmark/dataset_loaders/utils.py:174]: What is the diff between "pandas" and "dataframe"?

File: target_benchmark/retrievers/ottqa/OTTQARetriever.py [hunk 2]
```diff
@@ -32,18 +35,22 @@ def __init__(
         out_dir: str = default_out_dir,
         encoding: Literal["tfidf", "bm25"] = "tfidf",
         withtitle: bool = False,
+        ngram: int = 2,
```
Comment (madelonhulsebos) [inline target_benchmark/retrievers/ottqa/OTTQARetriever.py:38]: Is this a value chosen by ottqa?

**Review by madelonhulsebos — APPROVED** [review]

File: target_benchmark/dataset_loaders/AbsDatasetLoader.py [hunk 3]
```diff
@@ -145,23 +145,18 @@ def convert_corpus_table_to(
             df_tables = list(map(array_of_arrays_to_df, self.corpus[TABLE_COL_NAME]))
             converted_corpus[TABLE_COL_NAME] = df_tables
         elif in_memory_format == InMemoryDataFormat.DICTIONARY:
-            dict_tables = list(
-                map(array_of_arrays_to_dict, self.corpus[TABLE_COL_NAME])
-            )
+            dict_tables = list(map(array_of_arrays_to_dict, self.corpus[TABLE_COL_NAME]))
             converted_corpus[TABLE_COL_NAME] = dict_tables
+        count_tables = num_tables or self.num_tables
```
[the remark this answers is not in the document]
Comment (jixy2012) [reply target_benchmark/dataset_loaders/AbsDatasetLoader.py:150]: yes

File: target_benchmark/dataset_loaders/LoadersDataModels.py [hunk 4]
```diff
@@ -54,5 +49,10 @@ class HFDatasetConfigDataModel(DatasetConfigDataModel):
     )
 
 
+class NeedleInHaystackDatasetConfigDataModel(HFDatasetConfigDataModel):
+    query_type: str = "Needle in Haystack"
```
[the remark this answers is not in the document]
Comment (jixy2012) [reply target_benchmark/dataset_loaders/LoadersDataModels.py:53]: great catch thank you!! this is gonna be a bit annoying to maintain going forward if we add newer task types. i'll raise an issue for now and think of ways to better structure the code.

[the remark this answers is not in the document]
Comment (jixy2012) [reply target_benchmark/dataset_loaders/LoadersDataModels.py:53]: new issue [here](https://github.com/target-benchmark/target/issues/34)

File: target_benchmark/retrievers/ottqa/OTTQARetriever.py (hunk 2 above)
Comment (jixy2012) [reply target_benchmark/retrievers/ottqa/OTTQARetriever.py:38]: yes

File: target_benchmark/dataset_loaders/utils.py (hunk 1 above)
Comment (jixy2012) [reply target_benchmark/dataset_loaders/utils.py:174]: no difference, i think this code is from a while ago and i wrote both options as synonymous to converting to dataframe

File: target_benchmark/dataset_loaders/AbsDatasetLoader.py [hunk 5]
```diff
@@ -79,9 +82,7 @@ def _load_corpus(self) -> None:
     def _load_queries(self) -> None:
         pass
 
-    def persist_corpus_to(
-        self, format: Literal["csv", "json"], path: str = None
-    ) -> None:
+    def persist_corpus_to(self, format: Literal["csv", "json"], path: str = None) -> None:
```
[the remark this answers is not in the document]
Comment (jixy2012) [reply target_benchmark/dataset_loaders/AbsDatasetLoader.py:85]: could you elaborate on what is "parsing" the files? this original intention of this function was to allow users to save the corpus to a local file if they wanted to see what it looks like. it's not being used in the evaluation flow.

File: experiments/retrieval/utils.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    elif "tfidf_no_title" in retriever_name:
+        return OTTQARetriever(encoding="tfidf", withtitle=False)
+    elif "tfidf_with_title" in retriever_name:
+        return OTTQARetriever(encoding="tfidf", withtitle=True)
+    elif "bm25_no_title" in retriever_name:
+        return OTTQARetriever(encoding="bm25", withtitle=False)
+    elif "bm25_with_title" in retriever_name:
+        return OTTQARetriever(encoding="bm25", withtitle=True)
+    else:
+        raise ValueError(f"Passed in retriever {retriever_name} not yet supported")
```
[the remark this answers is not in the document]
Comment (jixy2012) [reply experiments/retrieval/utils.py:88]: indeed i haven't specified! it's a good idea to add a section for this in the readme and keep it updated as the newer retrievers get implemented

[the remark this answers is not in the document]
Comment (jixy2012) [reply experiments/retrieval/utils.py:88]: see new issue [here](https://github.com/target-benchmark/target/issues/35)

**Review by madelonhulsebos — APPROVED** [review]

**Pull request merged**